File: internal/git/branch.go
package git

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"os"
	"os/exec"
	"regexp"
	"strings"

	"github.com/rotisserie/eris"
)

// BranchInfo contains information about a branch
type BranchInfo struct {
	Name      string
	IsRemote  bool
	IsCurrent bool
}

var branchListSpecialChars = regexp.MustCompile(`[+*]`)

// ListLocalBranches lists all local branches in a repository
func ListLocalBranches(repoPath string) ([]string, error) {
	cmd := exec.Command("git", "-C", repoPath, "branch", "--format=%(refname:short)")
	output, err := cmd.Output()
	if err != nil {
		return nil, eris.Wrap(err, "failed to list local branches")
	}

	return parseGitBranchList(string(output)), nil
}

// ListRemoteBranches lists all remote branches in a repository
// Returns branch names without the "origin/" prefix (e.g., "main" instead of "origin/main")
// For bare repositories, lists branches from refs/heads/
func ListRemoteBranches(repoPath string) ([]string, error) {
	// Try listing branches using for-each-ref which works for both bare and normal repos
	cmd := exec.Command(
		"git",
		"-C",
		repoPath,
		"for-each-ref",
		"--format=%(refname:short)",
		"refs/heads/",
	)
	output, err := cmd.Output()
	if err != nil {
		// Fallback to branch -r for normal repos
		cmd = exec.Command("git", "-C", repoPath, "branch", "-r", "--format=%(refname:short)")
		output, err = cmd.Output()
		if err != nil {
			return nil, eris.Wrap(err, "failed to list remote branches")
		}

		branches := parseGitBranchList(string(output))

		// Remove "origin/" prefix and filter out HEAD
		var result []string
		for _, branch := range branches {
			if strings.Contains(branch, "HEAD") {
				continue
			}
			// Remove "origin/" prefix
			if trimmed, ok := strings.CutPrefix(branch, "origin/"); ok {
				result = append(result, trimmed)
			} else {
				result = append(result, branch)
			}
		}

		return result, nil
	}

	// For bare repos, branches are directly under refs/heads/
	return parseGitBranchList(string(output)), nil
}

// StreamRemoteBranches returns a reader that streams branch names and the cleanup function
// The reader will output one branch name per line as git produces them
// The caller must call cleanup() when done to ensure the process terminates
func StreamRemoteBranches(ctx context.Context, repoPath string) (io.ReadCloser, error) {
	// Use for-each-ref which works for both bare and normal repos

	cmd := exec.CommandContext(
		ctx,
		"git",
		"-C",
		repoPath,
		"ls-remote",
		"--branches",
		"--tags",
	)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, eris.Wrap(err, "failed to create stdout pipe")
	}

	if err := cmd.Start(); err != nil {
		return nil, eris.Wrap(err, "failed to start git command")
	}

	// Create a pipe that will transform the output
	reader, writer := io.Pipe()

	// Transform output in a goroutine
	go func() {
		ctx, cancel := context.WithCancelCause(ctx)
		defer cancel(nil)

		defer func() {
			if err := writer.Close(); err != nil {
				fmt.Fprintf(os.Stderr, "Error closing writer: %v\n", err)
			}
			if err := stdout.Close(); err != nil {
				fmt.Fprintf(os.Stderr, "Error closing stdout: %v\n", err)
			}
		}()

		local, err := ListLocalBranches(repoPath)
		if err != nil {
			cancel(eris.Wrap(err, "failed to list local branches"))
		}

		localSet := make(map[string]struct{})
		for _, branch := range local {
			localSet[branch] = struct{}{}
			fmt.Fprintln(writer, branch) //nolint:errcheck
		}

		// Wait for command to finish when done reading
		defer func() {
			var err error
			defer cancel(err)

			if err = cmd.Wait(); err != nil {
				fmt.Fprintf(
					os.Stderr,
					"Git command error: %s\n",
					eris.ToString(err, true),
				)
			}
		}()

		scanner := bufio.NewScanner(stdout)
		skip := true
		for scanner.Scan() {
			if ctx.Err() != nil {
				return // Context cancelled
			}

			if skip {
				skip = false
				continue
			}

			fields := strings.Fields(strings.TrimSpace(scanner.Text()))
			if len(fields) < 2 {
				continue
			}

			branch := strings.TrimPrefix(strings.TrimSpace(fields[1]), "refs/heads/")
			if branch != "" && !strings.Contains(branch, "HEAD") {
				// Remove "origin/" prefix if present
				if after, ok := strings.CutPrefix(branch, "origin/"); ok {
					branch = after
				}

				// this should not pose a concurrent access issue as all other writes are done before starting the goroutine.
				if _, exists := localSet[branch]; exists {
					continue // Skip local branches
				}

				fmt.Fprintln(writer, branch) //nolint:errcheck
			}
		}
	}()

	return reader, nil
}

// ListAllBranches lists both local and remote branches
// Returns a list of BranchInfo with details about each branch
func ListAllBranches(repoPath string) ([]BranchInfo, error) {
	cmd := exec.Command("git", "-C", repoPath, "branch", "-a", "--format=%(refname:short)")
	output, err := cmd.Output()
	if err != nil {
		return nil, eris.Wrap(err, "failed to list all branches")
	}

	// Get current branch
	currentBranch, _ := GetCurrentBranch(repoPath)

	branches := parseGitBranchList(string(output))
	var result []BranchInfo

	seen := make(map[string]bool)

	for _, branch := range branches {
		if strings.Contains(branch, "HEAD") {
			continue
		}

		isRemote := strings.HasPrefix(branch, "origin/")
		branchName := branch
		if isRemote {
			branchName = strings.TrimPrefix(branch, "origin/")
		}

		// Skip duplicates (local and remote with same name)
		if seen[branchName] {
			continue
		}
		seen[branchName] = true

		result = append(result, BranchInfo{
			Name:      branchName,
			IsRemote:  isRemote,
			IsCurrent: branchName == currentBranch,
		})
	}

	return result, nil
}

// DoesBranchExist checks if a branch exists in the repository
// For bare repositories (which sesh uses), all branches are at refs/heads/*
// Returns (exists, isRemote, error) - isRemote is always false for bare repos
func DoesBranchExist(repoPath, branch string) (bool, bool, error) {
	// Check if branch exists at refs/heads/<branch>
	// This works for both bare repos and regular repos with local branches
	cmd := exec.Command(
		"git",
		"-C",
		repoPath,
		"show-ref",
		"--verify",
		"--quiet",
		"refs/heads/"+branch,
	)

	err := cmd.Run()
	if err == nil {
		// Branch exists at refs/heads/<branch>
		return true, false, nil
	}

	// Check if it's a non-existent ref (exit code 1) or an actual error
	if exitErr, ok := err.(*exec.ExitError); ok {
		if exitErr.ExitCode() == 1 {
			// Branch doesn't exist
			return false, false, nil
		}
	}

	// Some other error occurred
	return false, false, eris.Wrap(err, "failed to check branch existence")
}

// GetCurrentBranch retrieves the current branch name in a git repository
func GetCurrentBranch(repoPath string) (string, error) {
	cmd := exec.Command("git", "-C", repoPath, "branch", "--show-current")
	output, err := cmd.Output()
	if err != nil {
		return "", eris.Wrap(err, "failed to get current branch")
	}
	branch := strings.TrimSpace(string(output))
	if branch == "" {
		return "(detached)", nil
	}
	return branch, nil
}

// parseGitBranchList parses the output of git branch commands
func parseGitBranchList(output string) []string {
	var branches []string
	scanner := bufio.NewScanner(strings.NewReader(output))
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			branches = append(branches, line)
		}
	}
	return branches
}

File: src/git/branch.rs
use crate::error::{Error, Result};
use git2::{BranchType, Repository};

/// Check if a branch exists locally
pub fn branch_exists_local(repo: &Repository, branch: &str) -> Result<bool> {
    let branch_ref = format!("refs/heads/{}", branch);
    match repo.find_reference(&branch_ref) {
        Ok(_) => Ok(true),
        Err(_) => Ok(false),
    }
}

/// Get the upstream branch for a local branch
pub fn get_upstream(repo: &Repository, branch: &str) -> Result<String> {
    let branch_ref = repo
        .find_branch(branch, BranchType::Local)
        .map_err(|e| Error::GitError {
            message: format!("Branch {} not found: {}", branch, e),
        })?;

    let upstream = branch_ref.upstream().map_err(|e| Error::GitError {
        message: format!("No upstream for branch {}: {}", branch, e),
    })?;

    upstream
        .name()
        .map_err(|_| Error::GitError {
            message: "Upstream branch name is invalid UTF-8".to_string(),
        })?
        .ok_or_else(|| Error::GitError {
            message: "Upstream branch has no name".to_string(),
        })
        .map(|s| s.to_string())
}

/// Check if a remote branch exists
pub fn remote_branch_exists(repo: &Repository, upstream: &str) -> Result<bool> {
    // Parse upstream (e.g., "origin/main")
    let parts: Vec<&str> = upstream.split('/').collect();
    if parts.len() != 2 {
        return Ok(false);
    }

    let remote_name = parts[0];
    let branch_name = parts[1];

    let remote_ref = format!("refs/remotes/{}/{}", remote_name, branch_name);
    match repo.find_reference(&remote_ref) {
        Ok(_) => Ok(true),
        Err(_) => Ok(false),
    }
}

/// List all local branches
pub fn list_local_branches(repo: &Repository) -> Result<Vec<String>> {
    let branches = repo
        .branches(Some(BranchType::Local))
        .map_err(|e| Error::GitError {
            message: format!("Failed to list branches: {}", e),
        })?;

    let mut branch_names = Vec::new();
    for branch in branches {
        let (branch, _) = branch.map_err(|e| Error::GitError {
            message: format!("Failed to process branch: {}", e),
        })?;

        if let Ok(Some(name)) = branch.name() {
            branch_names.push(name.to_string());
        }
    }

    Ok(branch_names)
}

/// List all remote branches
pub fn list_remote_branches(repo: &Repository) -> Result<Vec<String>> {
    let branches = repo
        .branches(Some(BranchType::Remote))
        .map_err(|e| Error::GitError {
            message: format!("Failed to list remote branches: {}", e),
        })?;

    let mut branch_names = Vec::new();
    for branch in branches {
        let (branch, _) = branch.map_err(|e| Error::GitError {
            message: format!("Failed to process branch: {}", e),
        })?;

        if let Ok(Some(name)) = branch.name() {
            // Remove "origin/" prefix
            let name = name
                .strip_prefix("origin/")
                .or_else(|| name.strip_prefix("remotes/origin/"))
                .unwrap_or(name);
            branch_names.push(name.to_string());
        }
    }

    Ok(branch_names)
}

/// List all branches (local and remote, deduplicated)
pub fn list_all_branches(repo: &Repository) -> Result<Vec<String>> {
    let mut branches = std::collections::HashSet::new();

    // Add local branches
    for branch in list_local_branches(repo)? {
        branches.insert(branch);
    }

    // Add remote branches
    for branch in list_remote_branches(repo)? {
        branches.insert(branch);
    }

    let mut result: Vec<String> = branches.into_iter().collect();
    result.sort();
    Ok(result)
}
